Article: case study for GDPR compliancy (#3)
pagopa/io-docs: docs/articles/case-study-gdpr-it.md

- [annullare la precedente
  richesta](https://github.com/pagopa/io-functions-app/tree/597853ffbb3b0e1f8594592fe04b302cad5fdee5/AbortUserDataProcessing).

### Workflow di cancellazione

Qui entrano in gioco le Durable Functions. Durable Functions è un'estensione di
Azure Functions che facilita lo sviluppo di workflow in cui si sovrappongo _long
running job_, interazioni umane, timer e operazioni asincrone in generale. Per
citare la documentazione ufficiale:

> Durable Functions è un'estensione di Funzioni di Azure che consente di
> scrivere funzioni con stato in un ambiente di calcolo serverless. L'estensione
> permette di definire flussi di lavoro con stato, scrivendo funzioni
> dell'agente di orchestrazione, ed entità con stato, scrivendo funzioni di
> entità tramite il modello di programmazione di Funzioni di Azure. Dietro le
> quinte, l'estensione gestisce automaticamente lo stato, i checkpoint e i
> riavvii, consentendo di concentrarsi sulla logica di business.

Per maggiori dettagli rimandiamo alla [documentazione di Azure Durable
Functions](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-overview).

> Noi del team di IO facciamo largo uso di queste funzionalità e in particolare

Reviewer: ```suggestion
Nell'implementazione del backend di IO facciamo largo uso di queste funzionalità e in particolare
```

> due tipologie di funzioni chiamate _Orchestrator_ e _Activity_. [La
> documentazione di
> Azure](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-types-features-overview#orchestrator-functions)
> è il miglior inizio per capire questi due strumenti, volendoli però riassumere
> potremmo dire che:

Reviewer: ```suggestion
due tipologie di funzioni: [_Orchestrator_ e _Activity_](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-types-features-overview#orchestrator-functions).
```

Tramite gli Orchestrator è possibile coordinare l'esecuzione di diverse
Activity. L'implementazione in Javascript è fatta utilizzando le [Generator
functions](https://developer.mozilla.org/it/docs/Web/JavaScript/Guida/Iteratori_e_generatori),
oggetti che meglio si addicono ad uno stile di programmazione imperativo anzichè
funzionale come invece saremmo portati a
[preferire](https://pagopa.github.io/io-docs/io-handbook/development-guidelines#general-guidelines);
compromesso che però accettiamo di buon grado vista la potenza e la versalità
che questo strumento ci mette a disposizione.

[Questo il workflow di
cancellazione](https://github.com/pagopa/io-functions-admin/blob/fa05bc96b6a756d4b8f14769a59b556d0709eb7a/UserDataDeleteOrchestrator/handler.ts#L320)
implementato tramite una Orchestrator function. É una funzione abbastanza lunga
(180 righe), ma sorprendentemente lineare in considerazione della logica articolata che implementa.
Come detto, lo stile è imperativo e le istruzioni vengono eseguite una dopo l'altra a meno di errori, che vengono gestiti nel `catch` finale. Ci sono però alcuni passaggi che meritano un approfondimento, di seguito proviamo a raccontarli.

> #### focus: gestione wait/interrupt

Reviewer: ```suggestion
#### Focus: gestione "periodo di grazia" per eliminazione contenuti utente
```

In pratica, interrompiamo l'esecuzione del workflow e decidiamo che possa essere riattivata solo in due casi: se il cittadino ci richiede l'annullamento o se sono passati 7 giorni. Alla riattivazione, controlliamo quale sia l'evento scatenante e in base a questo decidiamo con quale azione proseguire.

In pseudo-codice possiamo descriverlo così:

```
ATTENDI riattivazione CON ["richiesto annullamento" OPPURE "passati 7 giorni"]

SE riattivazione = "passati 7 giorni"
  > cancella i dati del cittadino
ALTRIMENTI
  > annulla la cancellazione
```

La codifica in Typescript usando la sdk fornita da Azure è abbastanza fedele allo pseudo-codice:

```ts
// timer a 7 giorni
const intervalExpiredEvent = context.df.createTimer(
  addDays(context.df.currentUtcDateTime, 7)
);
// listener per evento di annullamento
const canceledRequestEvent = context.df.waitForExternalEvent(ABORT_EVENT);

// il primo dei due eventi che succede
const triggeredEvent = yield context.df.Task.any([
  intervalExpiredEvent,
  canceledRequestEvent,
]);

if (triggeredEvent === intervalExpiredEvent) {
  // cancella i dati del cittadino
} else {
  // annulla la cancellazione
}
```

> La parte interessante è il metodo `Task.any` che risolve la _race condition_ tra gli _N_ eventi in input (due in questo caso), istruendo l'Orchestrator per bloccarsi e riattivarsi. L'implemetazione finale è consultabile in [questo punto](https://github.com/pagopa/io-functions-admin/blob/924a7ed8371681ff0fc8b4b2869695d0808ac9bf/UserDataDeleteOrchestrator%2Fhandler.ts#L366) del workflow.

Reviewer: ```suggestion
La parte interessante è il metodo `Task.any`: il metodo ritorna la prima attività conclusa tra quelle passate in input, nel caso specifico, l'evento di expire o la richiesta di cancellazione. L'implemetazione finale è consultabile in [questo punto](https://github.com/pagopa/io-functions-admin/blob/924a7ed8371681ff0fc8b4b2869695d0808ac9bf/UserDataDeleteOrchestrator%2Fhandler.ts#L366) del workflow.
```

### Azionamento del workflow

Piccola menzione per il meccanismo incaricato di azionare il workflow di cui abbiamo parlato. Attualmente viene gestito tramite un trigger sulla collection `user-data-collection` del database CosmosDB, richiamato ad ogni documento inserito. Guardando [l'implementazione della funzione](https://github.com/pagopa/io-functions-admin/blob/dad714275eb0bf505916ce375d390bbfe6360582/UserDataProcessingTrigger%2Findex.ts#L1), ci si accorge che per ogni documento si decide se:

* [avviare un workflow di accesso ai dati](https://github.com/pagopa/io-functions-admin/blob/dad714275eb0bf505916ce375d390bbfe6360582/UserDataProcessingTrigger%2Findex.ts#L116)
* [avviare un workflow di cancellazione](https://github.com/pagopa/io-functions-admin/blob/dad714275eb0bf505916ce375d390bbfe6360582/UserDataProcessingTrigger%2Findex.ts#L133)
* [emettere l'evento di annullamento della cancellazione](https://github.com/pagopa/io-functions-admin/blob/dad714275eb0bf505916ce375d390bbfe6360582/UserDataProcessingTrigger%2Findex.ts#L149)

o semplicemente ignorarlo, a seconda dei casi. 

> Questo trigger è stata una soluzione molto semplice da implementare grazie al [_binding_](https://github.com/pagopa/io-functions-admin/blob/a1bb2b16edff9c04bd86da6cffe405b097ddacaf/UserDataProcessingTrigger%2Ffunction.json#L3) offerto dal runtime Azure Functions. Potremmo pensare di sostituirlo con un sistema più sofisticato, soprattutto considerando che non è compatibile con in nostro processo di _blue/green deployment_, tuttavia per il momento si sta comportando molto bene.

Reviewer: ```suggestion
Questo trigger è stata una soluzione molto semplice da implementare grazie al [_binding_](https://github.com/pagopa/io-functions-admin/blob/a1bb2b16edff9c04bd86da6cffe405b097ddacaf/UserDataProcessingTrigger%2Ffunction.json#L3) offerto dal runtime Azure Functions.
```

non è chiaro cosa significa che è incompatibile, ometterei

---
title: 'Case study: garantire il diritto all''oblio'
created: '2020-11-13T16:43:39.086Z'
modified: '2020-11-14T09:47:45.393Z'
---

# Case study: garantire il diritto all'oblio

Quando progettiamo una nuova funzionalità per l'app IO, la più grande sfida che dobbiamo affrontare è quella di saper coniugare un'esperienza utente morbida con elevati requisiti di sicurezza, un'implementazione lineare con la complessità del contesto giuridico in cui IO si colloca rispetto al Cittadino. Questo richiede forte collaborazione tra sviluppatori, service designer, legali, esperti di sicurezza e di operations.

Questo processo interdisciplinare è stato messo particolarmente alla prova per recepire le norme del GDPR e integrarne i principi all'interno dell'app IO. In particolare, ai Cittadini deve essere garantito il _diritto all'oblio_, cioè deve essere rispettata la loro eventuale decisione di rimuovere i propri dati personali che vengono conservati dall'app IO.

> In questo articolo vogliamo condividere le soluzioni tecniche messe in campo per garantire il diritto all'oblio del Cittadino. In particolare vorremmo condividere l'utilizzo di Azure Durable Functions, e di come abbiamo tratto veramente grande beneficio da questo strumento.

Reviewer 2: `Cittadino` minuscolo?

Author: Secondo me no perché è un attore del sistema IO, quasi una keyword

Reviewer 2: In questa frase non è un attore, è un semplice cittadino, nell'elenco degli use case invece si ci sta che sia maiuscolo

## Specifiche

> L'implementazione che andremo a presentare è stata pensata per soddisfare i seguenti requisiti:

Reviewer 2: Andremo > andiamo?

* Come Cittadino voglio poter richiedere la cancellazione dei miei dati.
* Come Cittadino voglio poter ritirare la precedente richiesta entro 7gg.

> * Come DPO voglio che sia tenuta traccia di tutte le richieste del Cittadino e dello stato di lavorazione delle stesse.

Reviewer 2: ```suggestion
* Come DPO (Data Processing Officer) voglio che sia tenuta traccia di tutte le richieste del Cittadino e dello stato di elaborazione delle stesse.
```

> * Come Cittadino voglio essere notificato dell'avvenuta cancellazione

Reviewer 2: ```suggestion
* Come Cittadino voglio essere avvisato dell'avvenuta cancellazione
```

> * Come DPO voglio mantenere i dati del Cittadino per gestire eventuali contestazioni o contenziosi

Reviewer 2: Questo che significa? Che cancelliamo i dati per finta?

Reviewer: li spostiamo in uno storage che ha una durata prefissata (di modo che non sono più accessibili al CIT ma possano esser recuperati per i SAS che hanno inviato i messaggi, dovrebbe essere in DPIA)

Author: > li spostiamo in uno storage che ha una durata prefissata (di modo che non sono più accessibili al CIT ma possano esser recuperati per i SAS che hanno inviato i messaggi, dovrebbe essere in DPIA)

^^^ this. Poi valutiamo se specificare meglio questo dettaglio.

## Implementazione

> ### Task

Reviewer 2: Possiamo linkare le stories su pivotal?

Author: questi task sono una rielaborazione a posteriori, non abbiamo realmente lavorato in maniera così ben definita. Però sicuramente possiamo linkare la epic che raccoglie tutti i task

* Eliminare i dati utente
* Disiscrivere l'utente dai feed degli enti
* Inviare mail di completamento al cittadino

### Modellazione dei dati
Le informazioni delle richieste di cancellazione e dei suoi stati di lavorazione sono state modellate tramite la definizione `UserDataProcessing`. [Qui trovi la definizione completa](https://github.com/pagopa/io-functions-commons/blob/c46d77a5e5e8175d05d57dfc98cbac11c0661f7d/src/models/user_data_processing.ts#L43), ma ti basta sapere che:
* le richieste sono identificate dal codice fiscale del Cittadino e dal tipo di richiesta (`choice`, che può aver valore `DOWNLOAD` o `DELETE`);
* lo stato di lavorazione può essere `PENDING`, `WIP`, `ABORTED`, `FAILED` o `CLOSED`;
* i record non vengono aggiornati ma salvati sul database in modalità _append-only_. 

### Endpoint 
Abbiamo messo a disposizione delle API REST due endpoint dedicati:
* [richiedere la cancellazione](https://github.com/pagopa/io-functions-app/tree/597853ffbb3b0e1f8594592fe04b302cad5fdee5/UpsertUserDataProcessing);
* [annullare la precedente richesta](https://github.com/pagopa/io-functions-app/tree/597853ffbb3b0e1f8594592fe04b302cad5fdee5/AbortUserDataProcessing).

> Sono due tipici handler http di cui non approfondiremo le implementazioni.

Reviewer 2: ```suggestion
Sono due tipici handler HTTP REST di cui non approfondiremo le implementazioni.
```

### Workflow di cancellazione
> TODO: qui spiegazione in prosa del workflow

Qui entrano in gioco le Durable Functions. 
Durable Functions è un'estensione di Azure Functions che facilita lo sviluppo di workflow in cui si sovrappongo _long running job_, interazioni umane, timer e operazioni asincrone in generale. Per citare la documentazione ufficiale:
> Durable Functions è un'estensione di Funzioni di Azure che consente di scrivere funzioni con stato in un ambiente di calcolo serverless. L'estensione permette di definire flussi di lavoro con stato, scrivendo funzioni dell'agente di orchestrazione, ed entità con stato, scrivendo funzioni di entità tramite il modello di programmazione di Funzioni di Azure. Dietro le quinte, l'estensione gestisce automaticamente lo stato, i checkpoint e i riavvii, consentendo di concentrarsi sulla logica di business.

Per maggiori dettagli rimandiamo alla [documentazione di Azure Durable Functions](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-overview).

> Noi del team di IO facciamo largo uso di queste funzionalità e in particolare due tipo di funzioni chiamate _Orchestrator_ e _Activity_. [La documentazione di Azure](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-types-features-overview#orchestrator-functions) è il miglior inizio per capire questi due strumenti, volendoli però riassumere potremmo dire che:

Reviewer 2: ```suggestion
Noi del team di IO facciamo largo uso di queste funzionalità e in particolare due tipologie di funzioni chiamate _Orchestrator_ e _Activity_. [La documentazione di Azure](https://docs.microsoft.com/it-it/azure/azure-functions/durable/durable-functions-types-features-overview#orchestrator-functions) è il miglior inizio per capire questi due strumenti, volendoli però riassumere potremmo dire che:
```

* le Activity Functions sono dei singoli job, solitamente asincroni (letture/scritture su database, storage, interrogazione di API di terze parti, etc);
* le Orchestrator Functions sono delle istanze che mantengono un proprio stato e possono essere messe in pausa e riattivate.

> Come intuibile dal nome, tramite gli Orchestrator è possibile coordinare l'esecuzione di diverse Activity. L'implementazione in Javascript è fatta utilizzando le [Generator functions](https://developer.mozilla.org/it/docs/Web/JavaScript/Guida/Iteratori_e_generatori), oggetti che meglio si addicono ad uno stile di programmazione imperativo anzichè funzionale come invece saremmo portati a [preferire](https://pagopa.github.io/io-docs/io-handbook/development-guidelines#general-guidelines); compromesso che però accettiamo di buon grado vista la potenza e la versalità dello che questo strumento ci offre.

Reviewer 2: ```suggestion
Come intuibile dal nome, tramite gli Orchestrator è possibile coordinare l'esecuzione di diverse Activity. L'implementazione in Javascript è fatta utilizzando le [Generator functions](https://developer.mozilla.org/it/docs/Web/JavaScript/Guida/Iteratori_e_generatori), oggetti che meglio si addicono ad uno stile di programmazione imperativo anzichè funzionale come invece saremmo portati a [preferire](https://pagopa.github.io/io-docs/io-handbook/development-guidelines#general-guidelines); compromesso che però accettiamo di buon grado vista la potenza e la versalità che questo strumento ci mette a disposizione.
```

> [Questo il workflow di cancellazione](https://github.com/pagopa/io-functions-admin/blob/fa05bc96b6a756d4b8f14769a59b556d0709eb7a/UserDataDeleteOrchestrator/handler.ts#L320) implementato tramite una Orchestrator function. É una funziona abbastanza lunga (180 righe), ma sorprendentemente lineare in considerazione 

Reviewer 2: Frase sospesa?

Author: si si, è ancora draft
